File: lingting-http/src/main/java/live/lingting/framework/http/api/ApiClient.java
package live.lingting.framework.http.api;

import live.lingting.framework.http.HttpClient;
import live.lingting.framework.http.HttpMethod;
import live.lingting.framework.http.HttpRequest;
import live.lingting.framework.http.HttpResponse;
import live.lingting.framework.http.HttpUrlBuilder;
import live.lingting.framework.http.body.BodySource;
import live.lingting.framework.http.header.HttpHeaders;
import live.lingting.framework.value.multi.StringMultiValue;
import lombok.Getter;
import lombok.Setter;
import lombok.SneakyThrows;
import org.slf4j.Logger;

import java.net.URI;
import java.time.Duration;

/**
 * @author lingting 2024-09-14 15:33
 */
public abstract class ApiClient<R extends ApiRequest> {

	@Setter
	static HttpClient defaultClient = HttpClient.okhttp()
		.disableSsl()
		.timeout(Duration.ofSeconds(15), Duration.ofSeconds(30))
		.build();

	@Getter
	protected final Logger log = org.slf4j.LoggerFactory.getLogger(getClass());

	protected final String host;

	@Setter
	protected HttpClient client = defaultClient;

	protected ApiClient(String host) {
		this.host = host;
	}

	protected void customize(R request) {
		//
	}

	protected void customize(HttpHeaders headers) {
		//
	}

	protected void customize(R request, HttpHeaders headers) {
		//
	}

	protected void customize(HttpUrlBuilder builder) {
		//
	}

	protected void customize(R request, HttpRequest.Builder builder) {
		//
	}

	protected void customize(R request, HttpHeaders headers, BodySource source, StringMultiValue params) {
		//
	}

	protected abstract HttpResponse checkout(R request, HttpResponse response);

	@SneakyThrows
	protected HttpResponse call(R r) {
		r.onCall();
		customize(r);

		HttpMethod method = r.method();
		HttpHeaders headers = HttpHeaders.of(r.getHeaders());
		BodySource body = r.body();

		customize(headers);
		customize(r, headers);

		String path = r.path();
		r.onParams();
		HttpUrlBuilder urlBuilder = HttpUrlBuilder.builder().https().host(host).uri(path).addParams(r.getParams());
		customize(urlBuilder);

		HttpRequest.Builder builder = HttpRequest.builder();
		URI uri = urlBuilder.buildUri();
		builder.url(uri);
		headers.host(uri.getHost());

		customize(r, builder);
		customize(r, headers, body, urlBuilder.params());
		builder.headers(headers);
		builder.method(method.name()).body(body);

		HttpRequest request = builder.build();
		HttpResponse response = client.request(request);
		return checkout(r, response);
	}

}

File: lingting-http/src/main/kotlin/live/lingting/framework/http/api/ApiClient.kt
package live.lingting.framework.http.api

import live.lingting.framework.http.HttpClient
import live.lingting.framework.http.HttpRequest
import live.lingting.framework.http.HttpResponse
import live.lingting.framework.http.HttpUrlBuilder
import live.lingting.framework.http.body.BodySource
import live.lingting.framework.http.header.HttpHeaders
import live.lingting.framework.util.Slf4jUtils.logger
import java.time.Duration

/**
 * @author lingting 2024-09-14 15:33
 */
abstract class ApiClient<R : ApiRequest> @JvmOverloads protected constructor(
    @JvmField protected val host: String,
    @JvmField protected val ssl: Boolean = true,
) {

    companion object {

        @JvmStatic
        var defaultClient: HttpClient = HttpClient.builder()
            .disableSsl()
            .callTimeout(Duration.ofSeconds(10))
            .connectTimeout(Duration.ofSeconds(15))
            .readTimeout(Duration.ofSeconds(30))
            .build()

    }

    @JvmField
    val log = logger()

    @JvmField
    var client: HttpClient = defaultClient

    protected open fun customize(request: R) {
        //
    }

    protected open fun customize(body: BodySource) {
        //
    }

    protected open fun customize(headers: HttpHeaders) {
        //
    }

    protected open fun customize(body: BodySource, headers: HttpHeaders) {
        //
    }

    protected open fun customize(request: R, headers: HttpHeaders) {
        //
    }

    protected open fun customize(builder: HttpUrlBuilder) {
        //
    }

    protected open fun customize(request: R, headers: HttpHeaders, source: BodySource, url: HttpUrlBuilder) {
        //
    }

    protected abstract fun checkout(request: R, response: HttpResponse): HttpResponse

    protected open fun urlBuilder(): HttpUrlBuilder {
        return HttpUrlBuilder.builder().let {
            if (ssl) {
                it.https()
            } else {
                it.http()
            }
        }.host(host)
    }

    protected open fun call(r: R): HttpResponse {
        r.onCall()
        customize(r)

        val headers = HttpHeaders.of(r.headers)
        val body = r.body()

        customize(body)
        customize(headers)
        customize(r, headers)
        customize(body, headers)

        val path = r.path()
        r.onParams()
        val urlBuilder = urlBuilder().pathSegment(path).addParams(r.params)
        customize(urlBuilder)

        headers.host(urlBuilder.headerHost())

        customize(r, headers, body, urlBuilder)
        val response = call(urlBuilder, r, headers, body)
        return checkout(r, response)
    }

    protected open fun call(urlBuilder: HttpUrlBuilder, r: R, headers: HttpHeaders, body: BodySource): HttpResponse {
        val request = buildRequest(urlBuilder, headers, r, body)
        return call(r, request)
    }

    protected open fun call(r: R, request: HttpRequest): HttpResponse {
        return client.request(request)
    }

    protected open fun buildRequest(
        urlBuilder: HttpUrlBuilder,
        headers: HttpHeaders,
        r: R,
        body: BodySource
    ): HttpRequest {
        val builder = HttpRequest.builder()
        builder.url(urlBuilder)
        builder.headers(headers)
        builder.method(r.method().name).body(body)
        val request = builder.build()
        return request
    }

}
